plugin/pkg/shootstate/admission.go

```diff
@@ -0,0 +1,158 @@
+func (d *ValidateShootStateDeletion) AssignReadyFunc(f admission.ReadyFunc) {
+	d.readyFunc = f
+	d.SetReadyFunc(f)
+}
+
+// SetExternalCoreInformerFactory gets the garden core informer factory and adds it.
+func (d *ValidateShootStateDeletion) SetExternalCoreInformerFactory(f externalinformer.SharedInformerFactory) {
+	shootStateInformer := f.Core().V1alpha1().ShootStates()
+	d.shootStateLister = shootStateInformer.Lister()
+
+	readyFuncs = append(readyFuncs, shootStateInformer.Informer().HasSynced)
+}
+
+// SetInternalGardenClientset gets the garden clientset and adds it
+func (d *ValidateShootStateDeletion) SetInternalGardenClientset(c gardenclientset.Interface) {
+	d.gardenClient = c
+}
+
+func (d *ValidateShootStateDeletion) waitUntilReady(attrs admission.Attributes) error {
+	// Wait until the caches have been synced
+	if d.readyFunc == nil {
+		d.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+
+	if !d.WaitForReady() {
+		return admission.NewForbidden(attrs, errors.New("not yet ready to handle request"))
+	}
+
+	return nil
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (d *ValidateShootStateDeletion) ValidateInitialization() error {
+	if d.shootStateLister == nil {
+		return errors.New("missing ShootState lister")
+	}
+
+	return nil
+}
+
+var _ admission.ValidationInterface = &ValidateShootStateDeletion{}
+
+// Validate makes admissions decisions based on deletion confirmation annotation.
+func (d *ValidateShootStateDeletion) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	if a.GetKind().GroupKind() != core.Kind("ShootState") {
+		return nil
+	}
+
+	if err := d.waitUntilReady(a); err != nil {
+		return fmt.Errorf("Err while waiting for ready %v", err)
+	}
+
+	if a.GetName() == "" {
+		return d.validateDeleteCollection(ctx, a)
+	}
+
+	return d.validateDelete(ctx, a)
+}
+
+func (d *ValidateShootStateDeletion) validateDeleteCollection(ctx context.Context, attrs admission.Attributes) error {
+	shootStateList, err := d.shootStateLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, shootState := range shootStateList {
+		if err := d.validateDelete(ctx, d.createAdmissionAttributes(shootState, attrs)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *ValidateShootStateDeletion) validateDelete(ctx context.Context, attrs admission.Attributes) error {
+	_, err := d.gardenClient.Garden().Shoots(attrs.GetNamespace()).Get(attrs.GetName(), metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	return admission.NewForbidden(attrs, fmt.Errorf("Shoot %s in namespace %s still exists", attrs.GetName(), attrs.GetNamespace()))
+}
+
+func (d *ValidateShootStateDeletion) createAdmissionAttributes(obj metav1.Object, a admission.Attributes) admission.Attributes {
```

Comment:
The createAdmissionAttributes adds the object's name to the new attributes. Otherwise it's empty due to DeleteCollection .
Anyway what makes much more sense here is to just make the validateDelete method just take the name and namespace as arguments and not deal with attributes.

Reply:
need to set the object's name to the attributes so that it can be used in  the error returned from `admission.NewForbidden(attrs, ...)`